src/llvm-late-gc-lowering.cpp

```diff
@@ -1186,6 +1178,74 @@ static Value *GetPtrForNumber(State &S, unsigned Num, Instruction *InsertionPoin
     return Val;
 }
 
+static void AddInPredLiveOuts(BasicBlock *BB, BitVector &LiveIn, State &S)
+{
+    bool First = true;
+    std::set<BasicBlock *> Visited;
+    std::vector<BasicBlock *> WorkList;
+    WorkList.push_back(BB);
+    while (!WorkList.empty()) {
+        BB = &*WorkList.back();
+        WorkList.pop_back();
+        for (BasicBlock *Pred : predecessors(BB)) {
+            if (Visited.insert(Pred).second)
+                continue;
+            if (!S.BBStates[Pred].HasSafepoint) {
+                WorkList.push_back(Pred);
+                continue;
+            } else {
+                int LastSP = S.BBStates[Pred].Safepoints.front();
+                if (First) {
+                    LiveIn |= S.LiveSets[LastSP];
+                    First = false;
+                } else {
+                    LiveIn &= S.LiveSets[LastSP];
+                }
+            }
+        }
+    }
+}
+
+void LateLowerGCFrame::PlaceGCFrameStore(State &S, unsigned R, unsigned MinColorRoot, const std::vector<int> &Colors, Value *GCFrame, Instruction *InsertionPoint) {
+    Value *Val = GetPtrForNumber(S, R, InsertionPoint);
+    Value *args[1] = {
+        ConstantInt::get(T_int32, Colors[R]+MinColorRoot)
+    };
+    GetElementPtrInst *gep = GetElementPtrInst::Create(T_prjlvalue, GCFrame, makeArrayRef(args));
+    gep->insertBefore(InsertionPoint);
+    Val = MaybeExtractUnion(Val, InsertionPoint);
+    // Pointee types don't have semantics, so the optimizer is
+    // free to rewrite them if convenient. We need to change
+    // it back here for the store.
+    if (Val->getType() != T_prjlvalue)
+        Val = new BitCastInst(Val, T_prjlvalue, "", InsertionPoint);
+    new StoreInst(Val, gep, InsertionPoint);
+}
+
+void LateLowerGCFrame::PlaceGCFrameStores(Function &F, State &S, unsigned MinColorRoot, const std::vector<int> &Colors, Value *GCFrame)
+{
+    for (auto &BB : F) {
+        if (!S.BBStates[&BB].HasSafepoint) {
+            continue;
+        }
+        BitVector LiveIn;
+        AddInPredLiveOuts(&BB, LiveIn, S);
```

Comment:
This appears to be O(N^2) in the number of BB. Could this be converged once when creating the LiveSet instead of recomputing for each BB?